diff --git a/hexagonrpcd/apps_std.c b/hexagonrpcd/apps_std.c
index 06519f8..602d196 100644
--- a/hexagonrpcd/apps_std.c
+++ b/hexagonrpcd/apps_std.c
@@ -120,6 +120,40 @@ static uint32_t apps_std_fread(void *data,
 	return 0;
 }
 
+static uint32_t apps_std_fwrite(void *data,
+				const struct fastrpc_io_buffer *inbufs,
+				struct fastrpc_io_buffer *outbufs)
+{
+	struct apps_std_ctx *ctx = data;
+	const struct {
+		uint32_t fd;
+		uint32_t buf_len;
+	} *first_in = inbufs[0].p;
+	struct {
+		uint32_t written;
+		uint32_t is_eof;
+	} *first_out = outbufs[0].p;
+	ssize_t ret;
+
+	ret = hexagonfs_write(ctx->fds, first_in->fd,
+			      inbufs[1].s, inbufs[1].p);
+	if (ret < 0) {
+		fprintf(stderr, "Could not write file: %s\n", strerror(-ret));
+		return AEE_EFAILED;
+	}
+
+#ifdef HEXAGONRPC_VERBOSE
+	printf("write(%u, %u) -> %ld\n", first_in->fd,
+					first_in->buf_len,
+					ret);
+#endif
+
+	first_out->written = ret;
+	first_out->is_eof = first_out->written < first_in->buf_len;
+
+	return 0;
+}
+
 static uint32_t apps_std_fseek(void *data,
 			       const struct fastrpc_io_buffer *inbufs,
 			       struct fastrpc_io_buffer *outbufs)
@@ -156,7 +190,6 @@ static uint32_t apps_std_fopen_with_env(void *data,
 {
 	struct apps_std_ctx *ctx = data;
 	uint32_t *out = outbufs[0].p;
-	char rw_mode;
 	int dirfd, fd;
 
 	// The name and environment variable must also be NULL-terminated
@@ -165,12 +198,6 @@ static uint32_t apps_std_fopen_with_env(void *data,
 	 || ((const char *) inbufs[4].p)[inbufs[4].s - 1] != 0)
 		return AEE_EBADPARM;
 
-	rw_mode = ((const char *) inbufs[4].p)[0];
-	if (rw_mode == 'w' || rw_mode == 'a') {
-		fprintf(stderr, "Tried to open %s for writing\n",
-				(const char *) inbufs[3].p);
-		return AEE_EUNSUPPORTED;
-	}
 
 	if (!strcmp(inbufs[1].p, "ADSP_LIBRARY_PATH")) {
 		dirfd = ctx->adsp_library_dirfd;
@@ -419,7 +446,10 @@ static const struct fastrpc_function_impl apps_std_procs[] = {
 		.def = &apps_std_fread_def,
 		.impl = apps_std_fread,
 	},
-	{ .def = NULL, .impl = NULL, },
+	{
+		.def = &apps_std_fwrite_def,
+		.impl = apps_std_fwrite,
+	},
 	{ .def = NULL, .impl = NULL, },
 	{ .def = NULL, .impl = NULL, },
 	{ .def = NULL, .impl = NULL, },
diff --git a/hexagonrpcd/hexagonfs.c b/hexagonrpcd/hexagonfs.c
index bd9f353..122151e 100644
--- a/hexagonrpcd/hexagonfs.c
+++ b/hexagonrpcd/hexagonfs.c
@@ -244,6 +244,23 @@ ssize_t hexagonfs_read(struct hexagonfs_fd **fds, int fileno, size_t size, void
 	return fd->ops->read(fd, size, ptr);
 }
 
+ssize_t hexagonfs_write(struct hexagonfs_fd **fds, int fileno, size_t size, const void *ptr)
+{
+	struct hexagonfs_fd *fd;
+
+	if (fileno < 0 || fileno >= HEXAGONFS_MAX_FD)
+		return -EBADF;
+
+	fd = fds[fileno];
+	if (fd == NULL)
+		return -EBADF;
+
+	if (fd->ops->write == NULL)
+		return -ENOSYS;
+
+	return fd->ops->write(fd, size, ptr);
+}
+
 int hexagonfs_readdir(struct hexagonfs_fd **fds, int fileno, size_t ent_size, char *ent)
 {
 	struct hexagonfs_fd *fd;
diff --git a/hexagonrpcd/hexagonfs.h b/hexagonrpcd/hexagonfs.h
index 7682604..65ac531 100644
--- a/hexagonrpcd/hexagonfs.h
+++ b/hexagonrpcd/hexagonfs.h
@@ -41,6 +41,7 @@ struct hexagonfs_file_ops {
 	int (*readdir)(struct hexagonfs_fd *fd, size_t size, char *out);
 	ssize_t (*read)(struct hexagonfs_fd *fd, size_t size, void *ptr);
 	int (*stat)(struct hexagonfs_fd *fd, struct stat *stats);
+	ssize_t (*write)(struct hexagonfs_fd *fd, size_t size, const void *ptr);
 	int (*seek)(struct hexagonfs_fd *fd, off_t off, int whence);
 };
 
@@ -77,5 +78,6 @@ int hexagonfs_fstat(struct hexagonfs_fd **fds, int fileno, struct stat *stats);
 int hexagonfs_lseek(struct hexagonfs_fd **fds, int fileno, off_t pos, int whence);
 int hexagonfs_readdir(struct hexagonfs_fd **fds, int fileno, size_t size, char *name);
 ssize_t hexagonfs_read(struct hexagonfs_fd **fds, int fileno, size_t size, void *ptr);
+ssize_t hexagonfs_write(struct hexagonfs_fd **fds, int fileno, size_t size, const void *ptr);
 
 #endif
diff --git a/hexagonrpcd/hexagonfs_mapped.c b/hexagonrpcd/hexagonfs_mapped.c
index fd28c8c..fd9cf20 100644
--- a/hexagonrpcd/hexagonfs_mapped.c
+++ b/hexagonrpcd/hexagonfs_mapped.c
@@ -51,17 +51,16 @@ static int mapped_from_dirent(const void *dirent_data, bool dir, void **fd_data)
 {
 	struct mapped_ctx *ctx;
 	const char *name = dirent_data;
-	int flags = O_RDONLY;
+	int flags = dir ? (O_RDONLY | O_DIRECTORY) : (O_RDWR | O_CREAT);
 	int ret;
 
 	ctx = malloc(sizeof(struct mapped_ctx));
 	if (ctx == NULL)
 		return -ENOMEM;
 
-	if (dir)
-		flags |= O_DIRECTORY;
-
-	ctx->fd = open(name, flags);
+	ctx->fd = dir ? open(name, flags) : open(name, flags, 0664);
+	if (ctx->fd == -1 && !dir && errno == EISDIR)
+		ctx->fd = open(name, O_RDONLY | O_DIRECTORY);
 	if (ctx->fd == -1) {
 		ret = -errno;
 		goto err;
@@ -86,7 +85,7 @@ static int mapped_openat(struct hexagonfs_fd *dir,
 	struct mapped_ctx *dir_ctx = dir->data;
 	struct hexagonfs_fd *fd;
 	struct mapped_ctx *ctx;
-	int flags = O_RDONLY;
+	int flags = expect_dir ? (O_RDONLY | O_DIRECTORY) : (O_RDWR | O_CREAT);
 	int ret;
 
 	ctx = malloc(sizeof(struct mapped_ctx));
@@ -99,10 +98,10 @@ static int mapped_openat(struct hexagonfs_fd *dir,
 		goto err;
 	}
 
-	if (expect_dir)
-		flags |= O_DIRECTORY;
-
-	ctx->fd = openat(dir_ctx->fd, segment, flags);
+	ctx->fd = expect_dir ? openat(dir_ctx->fd, segment, flags)
+			     : openat(dir_ctx->fd, segment, flags, 0664);
+	if (ctx->fd == -1 && !expect_dir && errno == EISDIR)
+		ctx->fd = openat(dir_ctx->fd, segment, O_RDONLY | O_DIRECTORY);
 	if (ctx->fd == -1) {
 		ret = -errno;
 		goto err_free_fd;
@@ -138,6 +137,18 @@ static ssize_t mapped_read(struct hexagonfs_fd *fd, size_t size, void *out)
 	return ret;
 }
 
+static ssize_t mapped_write(struct hexagonfs_fd *fd, size_t size, const void *in)
+{
+	struct mapped_ctx *ctx = fd->data;
+	ssize_t ret;
+
+	ret = write(ctx->fd, in, size);
+	if (ret < 0)
+		return -errno;
+
+	return ret;
+}
+
 static int mapped_readdir(struct hexagonfs_fd *fd, size_t size, char *out)
 {
 	struct mapped_ctx *ctx = fd->data;
@@ -253,6 +264,15 @@ static ssize_t mapped_or_empty_read(struct hexagonfs_fd *fd,
 	return 0;
 }
 
+static ssize_t mapped_or_empty_write(struct hexagonfs_fd *fd,
+				     size_t size, const void *in)
+{
+	if (fd->data)
+		return mapped_write(fd, size, in);
+
+	return 0;
+}
+
 static int mapped_or_empty_readdir(struct hexagonfs_fd *fd,
 				   size_t size, char *out)
 {
@@ -329,6 +349,7 @@ struct hexagonfs_file_ops hexagonfs_mapped_ops = {
 	.read = mapped_read,
 	.readdir = mapped_readdir,
 	.seek = mapped_seek,
+	.write = mapped_write,
 	.stat = mapped_stat,
 };
 
@@ -339,6 +360,7 @@ struct hexagonfs_file_ops hexagonfs_mapped_or_empty_ops = {
 	.read = mapped_or_empty_read,
 	.readdir = mapped_or_empty_readdir,
 	.seek = mapped_or_empty_seek,
+	.write = mapped_or_empty_write,
 	.stat = mapped_or_empty_stat,
 };
 
@@ -349,5 +371,6 @@ struct hexagonfs_file_ops hexagonfs_mapped_sysfs_ops = {
 	.read = mapped_read,
 	.readdir = mapped_readdir,
 	.seek = mapped_seek,
+	.write = mapped_write,
 	.stat = mapped_sysfs_stat,
 };
diff --git a/hexagonrpcd/interfaces/apps_std.def b/hexagonrpcd/interfaces/apps_std.def
index 2fbbdb1..35441e1 100644
--- a/hexagonrpcd/interfaces/apps_std.def
+++ b/hexagonrpcd/interfaces/apps_std.def
@@ -28,6 +28,7 @@ HEXAGONRPC_DEFINE_REMOTE_METHOD(1, apps_std_freopen, 1, 0, 1, 1)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(2, apps_std_fflush, 8, 0, 0, 0)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(3, apps_std_fclose, 1, 0, 0, 0)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(4, apps_std_fread, 1, 0, 2, 1)
+HEXAGONRPC_DEFINE_REMOTE_METHOD(5, apps_std_fwrite, 1, 1, 2, 0)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(9, apps_std_fseek, 3, 0, 0, 0)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(19, apps_std_fopen_with_env, 0, 4, 1, 0)
 HEXAGONRPC_DEFINE_REMOTE_METHOD(26, apps_std_opendir, 0, 1, 2, 0)
diff --git a/hexagonrpcd/rpcd_builder.c b/hexagonrpcd/rpcd_builder.c
index d374958..afb8b06 100644
--- a/hexagonrpcd/rpcd_builder.c
+++ b/hexagonrpcd/rpcd_builder.c
@@ -162,9 +162,7 @@ struct hexagonfs_dirent *construct_root_dir(const char *prefix, const char *dsp)
 	 */
 	persist_dir = hfs_mkdir("persist", 1,
 				hfs_mkdir("sensors", 1,
-					hfs_mkdir("registry", 1,
-						hfs_map("registry", sns_reg)
-					)
+					hfs_map_or_empty("registry", sns_reg)
 				)
 		      );
 
