diff -ruN iio-sensor-proxy-3.9/data/80-iio-sensor-proxy-libssc.rules patched/data/80-iio-sensor-proxy-libssc.rules
--- iio-sensor-proxy-3.9/data/80-iio-sensor-proxy-libssc.rules	1970-01-01 02:00:00.000000000 +0200
+++ patched/data/80-iio-sensor-proxy-libssc.rules	2026-07-01 21:25:53.674237539 +0300
@@ -0,0 +1,11 @@
+# iio-sensor-proxy: libssc
+# IIO sensor to D-Bus proxy with libssc support
+
+ACTION=="remove", GOTO="iio_sensor_proxy_libssc_end"
+
+# Set the sensor type for all the types we recognise for SDSP/ADSP.
+# Some devices expose the sensors via a separate SDSP while others re-use ADSP.
+# These DSPs expose itself as /dev/fastrpc-* in the misc subsystem.
+SUBSYSTEM=="misc", KERNEL=="fastrpc-*", ENV{IIO_SENSOR_PROXY_TYPE}+="ssc-accel ssc-light ssc-proximity ssc-compass"
+
+LABEL="iio_sensor_proxy_libssc_end"
diff -ruN iio-sensor-proxy-3.9/data/meson.build patched/data/meson.build
--- iio-sensor-proxy-3.9/data/meson.build	2026-03-02 21:19:51.000000000 +0200
+++ patched/data/meson.build	2026-07-01 21:25:53.675237475 +0300
@@ -2,8 +2,12 @@
   '80-iio-sensor-proxy.rules'
 )
 
+if get_option('ssc-support').enabled()
+  rules += files('80-iio-sensor-proxy-libssc.rules')
+endif
+
 install_data(
-  '80-iio-sensor-proxy.rules',
+  rules,
   install_dir: udev_rules_dir
 )
 
