diff --git src/drv-ssc-accel.c b/src/drv-ssc-accel.c
--- a/src/drv-ssc-accel.c	2026-03-02 21:19:51.000000000 +0200
+++ b/src/drv-ssc-accel.c	2026-07-02 02:48:22.567795729 +0300
@@ -30,20 +30,11 @@
 static gboolean
 ssc_accelerometer_discover (GUdevDevice *device)
 {
-	g_autoptr(SSCSensorAccelerometer) sensor = NULL;
 
 	/* Verify presence of FastRPC device */
 	if (!drv_check_udev_sensor_type (device, "ssc-accel", NULL))
 		return FALSE;
 
-	/* Open and close SSC accelerometer for discovering */
-	sensor = ssc_sensor_accelerometer_new_sync (NULL, NULL);
-	if (!sensor)
-		return FALSE;
-
-	if (!ssc_sensor_accelerometer_close_sync (sensor, NULL, NULL))
-		return FALSE;
-
 	g_debug ("Found SSC accelerometer at %s", g_udev_device_get_sysfs_path (device));
 	return TRUE;
 }
@@ -83,12 +74,15 @@
 
 	drv_data = (DrvData *) sensor_device->priv;
 	drv_data->sensor = ssc_sensor_accelerometer_new_sync (NULL, &error);
-	if (!drv_data->sensor)
+	if (!drv_data->sensor) {
 		g_warning ("Creating SSC accelerometer sensor failed: %s", error->message);
-	else
-		g_object_get (drv_data->sensor,
-		              SSC_SENSOR_NAME, &sensor_device->name,
-		              NULL);
+		g_free (drv_data);
+		g_free (sensor_device);
+		return NULL;
+	}
+	g_object_get (drv_data->sensor,
+	              SSC_SENSOR_NAME, &sensor_device->name,
+	              NULL);
 
 	/* Setup accel attributes */
 	drv_data->mount_matrix = setup_mount_matrix (device);
@@ -107,6 +101,9 @@
 		if (drv_data->measurement_id)
 			return;
 
+		if (!drv_data->sensor)
+			return;
+
 		/* Start listening for measurements */
 		drv_data->measurement_id = g_signal_connect (drv_data->sensor,
 					                     "measurement",
diff --git src/drv-ssc-compass.c b/src/drv-ssc-compass.c
--- a/src/drv-ssc-compass.c	2026-03-02 21:19:51.000000000 +0200
+++ b/src/drv-ssc-compass.c	2026-07-02 02:48:22.568944604 +0300
@@ -26,20 +26,11 @@
 static gboolean
 ssc_compass_discover (GUdevDevice *device)
 {
-	g_autoptr(SSCSensorCompass) sensor = NULL;
 
 	/* Verify presence of FastRPC device */
 	if (!drv_check_udev_sensor_type (device, "ssc-compass", NULL))
 		return FALSE;
 
-	/* Open and close SSC compass for discovering */
-	sensor = ssc_sensor_compass_new_sync (NULL, NULL);
-	if (!sensor)
-		return FALSE;
-
-	if (!ssc_sensor_compass_close_sync (sensor, NULL, NULL))
-		return FALSE;
-
 	g_debug ("Found SSC compass at %s", g_udev_device_get_sysfs_path (device));
 	return TRUE;
 }
@@ -67,13 +58,15 @@
 
 	drv_data = (DrvData *) sensor_device->priv;
 	drv_data->sensor = ssc_sensor_compass_new_sync (NULL, &error);
-	if (!drv_data->sensor)
+	if (!drv_data->sensor) {
 		g_warning ("Creating SSC compass sensor failed: %s", error->message);
-	else {
-	  g_object_get (drv_data->sensor,
-			SSC_SENSOR_NAME, &sensor_device->name,
-			NULL);
+		g_free (drv_data);
+		g_free (sensor_device);
+		return NULL;
 	}
+	g_object_get (drv_data->sensor,
+	              SSC_SENSOR_NAME, &sensor_device->name,
+	              NULL);
 
 	return sensor_device;
 }
diff --git src/drv-ssc-light.c b/src/drv-ssc-light.c
--- a/src/drv-ssc-light.c	2026-03-02 21:19:51.000000000 +0200
+++ b/src/drv-ssc-light.c	2026-07-02 02:48:22.569502681 +0300
@@ -26,20 +26,11 @@
 static gboolean
 ssc_light_discover (GUdevDevice *device)
 {
-	g_autoptr(SSCSensorLight) sensor = NULL;
 
 	/* Verify presence of FastRPC device */
 	if (!drv_check_udev_sensor_type (device, "ssc-light", NULL))
 		return FALSE;
 
-	/* Open and close SSC light sensor for discovering */
-	sensor = ssc_sensor_light_new_sync (NULL, NULL);
-	if (!sensor)
-		return FALSE;
-
-	if (!ssc_sensor_light_close_sync (sensor, NULL, NULL))
-		return FALSE;
-
 	g_debug ("Found SSC light at %s", g_udev_device_get_sysfs_path (device));
 	return TRUE;
 }
@@ -67,12 +58,15 @@
 
 	drv_data = (DrvData *) sensor_device->priv;
 	drv_data->sensor = ssc_sensor_light_new_sync (NULL, &error);
-	if (!drv_data->sensor)
+	if (!drv_data->sensor) {
 		g_warning ("Creating SSC light sensor failed: %s", error->message);
-	else
-		g_object_get (drv_data->sensor,
-		              SSC_SENSOR_NAME, &sensor_device->name,
-		              NULL);
+		g_free (drv_data);
+		g_free (sensor_device);
+		return NULL;
+	}
+	g_object_get (drv_data->sensor,
+	              SSC_SENSOR_NAME, &sensor_device->name,
+	              NULL);
 
 	return sensor_device;
 }
diff --git src/drv-ssc-proximity.c b/src/drv-ssc-proximity.c
--- a/src/drv-ssc-proximity.c	2026-03-02 21:19:51.000000000 +0200
+++ b/src/drv-ssc-proximity.c	2026-07-02 02:48:22.569596081 +0300
@@ -26,24 +26,11 @@
 static gboolean
 ssc_proximity_discover (GUdevDevice *device)
 {
-	SSCSensorProximity *sensor = NULL;
 
 	/* Verify presence of FastRPC device */
 	if (!drv_check_udev_sensor_type (device, "ssc-proximity", NULL))
 		return FALSE;
 
-	/* Open and close SSC proximity sensor for discovering */
-	sensor = ssc_sensor_proximity_new_sync (NULL, NULL);
-	if (!sensor)
-		return FALSE;
-
-	if (!ssc_sensor_proximity_close_sync (sensor, NULL, NULL)) {
-		g_clear_object (&sensor);
-		return FALSE;
-	}
-
-	g_clear_object (&sensor);
-
 	g_debug ("Found SSC proximity at %s", g_udev_device_get_sysfs_path (device));
 	return TRUE;
 }
@@ -69,12 +56,15 @@
 	sensor_device->priv = g_new0 (DrvData, 1);
 	drv_data = (DrvData *) sensor_device->priv;
 	drv_data->sensor = ssc_sensor_proximity_new_sync (NULL, &error);
-	if (!drv_data->sensor)
+	if (!drv_data->sensor) {
 		g_warning ("Creating SSC proximity sensor failed: %s", error->message);
-	else
-		g_object_get (drv_data->sensor,
-		              SSC_SENSOR_NAME, &sensor_device->name,
-		              NULL);
+		g_free (drv_data);
+		g_free (sensor_device);
+		return NULL;
+	}
+	g_object_get (drv_data->sensor,
+	              SSC_SENSOR_NAME, &sensor_device->name,
+	              NULL);
 
 	return sensor_device;
 }
@@ -112,8 +102,14 @@
 static void
 ssc_proximity_close (SensorDevice *sensor_device)
 {
+	g_autoptr(GError) error = NULL;
 	DrvData *drv_data = (DrvData *) sensor_device->priv;
 
+	if (drv_data->sensor) {
+		g_clear_signal_handler (&drv_data->measurement_id, drv_data->sensor);
+		if (!ssc_sensor_proximity_close_sync (drv_data->sensor, NULL, &error))
+			g_warning ("Closing SSC proximity sensor failed: %s", error->message);
+	}
 	g_clear_object (&drv_data->sensor);
 	g_clear_pointer (&sensor_device->priv, g_free);
 	g_free (sensor_device);
