diff --git a/arch/arm64/boot/dts/qcom/sm8250-xiaomi-pipa-common.dtsi b/arch/arm64/boot/dts/qcom/sm8250-xiaomi-pipa-common.dtsi
--- a/arch/arm64/boot/dts/qcom/sm8250-xiaomi-pipa-common.dtsi
+++ b/arch/arm64/boot/dts/qcom/sm8250-xiaomi-pipa-common.dtsi
@@ -528,7 +528,14 @@
 regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
 		};
 
---/* L2 and L4 are unused. */
+--/* L2 is unused. */
+
+--vreg_l4c_1p8: ldo4 {
+---regulator-name = "vreg_l4c_1p8";
+---regulator-min-microvolt = <1800000>;
+---regulator-max-microvolt = <1980000>;
+---regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+--};
 
 		vreg_l3c_1p2: ldo3 {
 			regulator-name = "vreg_l3c_1p2";
@@ -710,6 +717,25 @@
 	};
 };
 
+&i2c2 {
+-status = "okay";
+
+-nanosic_803: nanosic803@4c {
+--compatible = "nanosic,803";
+--status = "okay";
+--reg = <0x4c>;
+--pinctrl-names = "default";
+--pinctrl-0 = <&nanosic_default>;
+--reset-gpios = <&tlmm 141 GPIO_ACTIVE_LOW>;
+--sleep-gpios = <&tlmm 155 GPIO_ACTIVE_LOW>;
+--vdd-gpios = <&tlmm 127 GPIO_ACTIVE_HIGH>;
+--irq-gpios = <&tlmm 100 IRQ_TYPE_EDGE_RISING>;
+--touchpad-resolution-x = <2879>;
+--touchpad-resolution-y = <1799>;
+--vdd_1v8-supply = <&vreg_l4c_1p8>;
+--vdd_3v3-supply = <&vreg_l10c_3p3>;
+-};
+};
+
 &i2c3 {
 	clock-frequency = <400000>;
 	status = "okay";
@@ -1256,6 +1282,20 @@
 		bias-pull-up;
 		input-enable;
 	};
+
+-nanosic_default: nanosic-default-state {
+--sleep-pins {
+---pins = "gpio155";
+---function = "gpio";
+---drive-strength = <2>;
+---output-low;
+--};
+
+--reset-pins {
+---pins = "gpio141";
+---function = "gpio";
+---drive-strength = <2>;
+---output-low;
+--};
+-};
 
 	bt_en_state: bt-default-state {
 		bt-en-pins {
