--- a/drivers/input/keyboard/nanosic_803.c
+++ b/drivers/input/keyboard/nanosic_803.c
@@ -23,6 +23,7 @@
 #include <linux/debugfs.h>
 #include <linux/uaccess.h>
 #include <linux/ctype.h>
+#include <linux/hex.h>
 
 #define I2C_DATA_LENGTH_READ (68)
 #define I2C_DATA_LENGTH_WRITE (66)
@@ -746,7 +747,7 @@
 
 static void nanosic_touch_timer_callback(struct timer_list *t)
 {
-	struct nanosic_803_priv *nanosic_dev = from_timer(nanosic_dev, t, finger_timer);
+	struct nanosic_803_priv *nanosic_dev = timer_container_of(nanosic_dev, t, finger_timer);
 
 	if (nanosic_dev->finger_down) {
 		for (int i = 0; i<3; i++) {
