diff --git a/src/iio-sensor-proxy.c b/src/iio-sensor-proxy.c
index 2f0111c..d650b97 100644
--- a/src/iio-sensor-proxy.c
+++ b/src/iio-sensor-proxy.c
@@ -449,7 +449,8 @@ client_vanished_cb (GDBusConnection *connection,
 		guint watch_id;
 
 		ht = data->clients[i];
-		g_assert (ht);
+		if (ht == NULL)
+			continue;
 
 		watch_id = GPOINTER_TO_UINT (g_hash_table_lookup (ht, sender));
 		if (watch_id > 0)
@@ -750,6 +751,12 @@ name_acquired_handler (GDBusConnection *connection,
 	guint i;
 
 	data->client = g_udev_client_new (subsystems);
+
+	for (i = 0; i < NUM_SENSOR_TYPES; i++) {
+		data->clients[i] = create_clients_hash_table ();
+		data->sensor_startup_dbus_invocations_delayed[i] = g_ptr_array_new ();
+	}
+
 	if (!find_sensors (data->client, data))
 		goto bail;
 
@@ -759,9 +766,6 @@ name_acquired_handler (GDBusConnection *connection,
 	for (i = 0; i < NUM_SENSOR_TYPES; i++) {
 		SensorDevice *sensor_device;
 
-		data->clients[i] = create_clients_hash_table ();
-		data->sensor_startup_dbus_invocations_delayed[i] = g_ptr_array_new ();
-
 		if (!driver_type_exists (data, i))
 			continue;
 
