diff --git a/sound/soc/codecs/aw88261.c b/sound/soc/codecs/aw88261.c
index 9c3b229c854b..25f54987fcee 100644
--- a/sound/soc/codecs/aw88261.c
+++ b/sound/soc/codecs/aw88261.c
@@ -172,7 +172,7 @@ static int aw88261_dev_get_iis_status(struct aw_device *aw_dev)
 		if (pll_lock && clks_available)
 			return 0;
 
-		dev_err(aw_dev->dev,
+		dev_dbg(aw_dev->dev,
 			"IIS signal is not OK, pll_lock:%d, clks_available:%d, reg_val:0x%04x",
 			pll_lock, clks_available, reg_val);
 		return -EINVAL;
@@ -193,7 +193,7 @@ static int aw88261_dev_check_mode1_pll(struct aw_device *aw_dev)
 	for (i = 0; i < AW88261_DEV_SYSST_CHECK_MAX; i++) {
 		ret = aw88261_dev_get_iis_status(aw_dev);
 		if (ret) {
-			dev_err(aw_dev->dev, "mode1 iis signal check error");
+			dev_dbg(aw_dev->dev, "mode1 iis signal check error");
 			usleep_range(AW88261_2000_US, AW88261_2000_US + 10);
 		} else {
 			return ret;
@@ -227,7 +227,7 @@ static int aw88261_dev_check_mode2_pll(struct aw_device *aw_dev)
 	for (i = 0; i < AW88261_DEV_SYSST_CHECK_MAX; i++) {
 		ret = aw88261_dev_get_iis_status(aw_dev);
 		if (ret) {
-			dev_err(aw_dev->dev, "mode2 iis signal check error");
+			dev_dbg(aw_dev->dev, "mode2 iis signal check error");
 			usleep_range(AW88261_2000_US, AW88261_2000_US + 10);
 		} else {
 			break;
@@ -242,7 +242,7 @@ static int aw88261_dev_check_mode2_pll(struct aw_device *aw_dev)
 		for (i = 0; i < AW88261_DEV_SYSST_CHECK_MAX; i++) {
 			ret = aw88261_dev_check_mode1_pll(aw_dev);
 			if (ret) {
-				dev_err(aw_dev->dev, "mode2 switch to mode1, iis signal check error");
+				dev_dbg(aw_dev->dev, "mode2 switch to mode1, iis signal check error");
 				usleep_range(AW88261_2000_US, AW88261_2000_US + 10);
 			} else {
 				break;
@@ -265,6 +265,7 @@ static int aw88261_dev_check_syspll(struct aw_device *aw_dev)
 			ret = aw88261_dev_check_mode1_pll(aw_dev);
 			if (!ret)
 				break;
+			usleep_range(50000, 60000);
 			retry_count++;
 		}
 
@@ -305,7 +306,7 @@ static int aw88261_dev_check_sysst(struct aw_device *aw_dev)
 		check_val = reg_val & (~AW88261_BIT_SYSST_CHECK_MASK)
 							& AW88261_BIT_SYSST_CHECK;
 		if (check_val != AW88261_BIT_SYSST_CHECK) {
-			dev_err(aw_dev->dev, "check sysst fail, reg_val=0x%04x, check:0x%x",
+			dev_dbg(aw_dev->dev, "check sysst fail, reg_val=0x%04x, check:0x%x",
 				reg_val, AW88261_BIT_SYSST_CHECK);
 			usleep_range(AW88261_2000_US, AW88261_2000_US + 10);
 		} else {
@@ -773,10 +774,10 @@ static void aw88261_start_pa(struct aw88261 *aw88261)
 		}
 		ret = aw88261_dev_start(aw88261);
 		if (ret) {
-			dev_err(aw88261->aw_pa->dev, "aw88261 device start failed. retry = %d", i);
+			dev_warn(aw88261->aw_pa->dev, "aw88261 device start failed. retry = %d", i);
 			continue;
 		} else {
-			dev_info(aw88261->aw_pa->dev, "start success\n");
+			dev_dbg(aw88261->aw_pa->dev, "start success\n");
 			break;
 		}
 	}
@@ -1099,7 +1100,7 @@ static int aw88261_playback_event(struct snd_soc_dapm_widget *w,
 	mutex_lock(&aw88261->lock);
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
-		aw88261_start(aw88261, AW88261_ASYNC_START);
+		aw88261_start(aw88261, AW88261_SYNC_START);
 		break;
 	case SND_SOC_DAPM_POST_PMD:
 		aw88261_dev_stop(aw88261->aw_pa);
