From d2596103608f35f1172a83e70e1726d8b555e31f Mon Sep 17 00:00:00 2001
From: Ayman Ragab <aymanrar2c@gmail.com>
Date: Sun, 19 Jul 2026 14:15:58 +0300
Subject: [PATCH 8/8] media: i2c: ov13b10: add get_selection pad operation

---
 drivers/media/i2c/ov13b10.c | 25 +++++++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/media/i2c/ov13b10.c b/drivers/media/i2c/ov13b10.c
index 18bb63f96262..7b5d07172cfe 100644
--- a/drivers/media/i2c/ov13b10.c
+++ b/drivers/media/i2c/ov13b10.c
@@ -1165,6 +1165,30 @@ ov13b10_set_pad_format(struct v4l2_subdev *sd,
 	return 0;
 }
 
+static int ov13b10_get_selection(struct v4l2_subdev *sd,
+				 struct v4l2_subdev_state *sd_state,
+				 struct v4l2_subdev_selection *sel)
+{
+	switch (sel->target) {
+	case V4L2_SEL_TGT_CROP_DEFAULT:
+	case V4L2_SEL_TGT_CROP_BOUNDS:
+	case V4L2_SEL_TGT_NATIVE_SIZE:
+		sel->r.left = 0;
+		sel->r.top = 0;
+		sel->r.width = 4208;
+		sel->r.height = 3120;
+		return 0;
+	case V4L2_SEL_TGT_CROP:
+		if (sel->which == V4L2_SUBDEV_FORMAT_TRY)
+			sel->r = *v4l2_subdev_state_get_crop(sd_state, sel->pad);
+		else
+			sel->r = (struct v4l2_rect){0, 0, 4208, 3120};
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
 /* Verify chip ID */
 static int ov13b10_identify_module(struct ov13b10 *ov13b)
 {
@@ -1372,6 +1396,7 @@ static const struct v4l2_subdev_pad_ops ov13b10_pad_ops = {
 	.enum_mbus_code = ov13b10_enum_mbus_code,
 	.get_fmt = ov13b10_get_pad_format,
 	.set_fmt = ov13b10_set_pad_format,
+	.get_selection = ov13b10_get_selection,
 	.enum_frame_size = ov13b10_enum_frame_size,
 };
 
-- 
2.54.0

